File: Codemania_Classical_1.0/D. Deserted Freshers/sol-dfs.cpp
#include <algorithm>
#include <functional>
#include <iostream>
#include <numeric>
#include <vector>

using namespace std;

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);
    cout.tie(nullptr);

    int n, m;
    cin >> n >> m;

    vector<int> a(n), visited(n);
    vector<vector<int>> g(n), components(n);

    for (auto &x : a) cin >> x;

    for (int i = 0; i < m; ++i) {
        int u, v;
        cin >> u >> v;
        --u, --v;
        g[u].push_back(v);
        g[v].push_back(u);
    }

    function<void(int, int)> dfs = [&](int v, int component) {
        visited[v] = true;
        components[component].push_back(v);
        for (auto &u : g[v]) {
            if (!visited[u]) {
                dfs(u, component);
            }
        }
    };

    int total_components = 0;
    for (int i = 0; i < n; ++i) {
        if (!visited[i]) {
            dfs(i, total_components);
            total_components++;
        }
    }

    vector<int> indices(n);

    iota(indices.begin(), indices.end(), 0);

    sort(indices.begin(), indices.end(),
         [&](const int i, const int j) { return a[i] > a[j]; });

    sort(components.begin(), components.begin() + total_components,
         [](const vector<int> &i, const vector<int> &j) {
             return i.size() > j.size();
         });

    int current_rank = 0;
    vector<int> assigned_vertex(n);
    for (int component = 0; component < total_components; ++component) {
        for (auto &x : components[component]) {
            assigned_vertex[indices[current_rank]] = x;
            current_rank++;
        }
    }

    for (auto &x : assigned_vertex) cout << x + 1 << ' ';

    cout << '\n';

    return 0;
}

File: IITD/Codemania_Classical_1.0/D. Deserted Freshers/sol-dfs.py
# Fast IO

import sys
input = sys.stdin.readline
def print(x, end='\n'):
    sys.stdout.write(str(x) + end)

# IO helpers

def get_int():
    return int(input())
def get_list_ints():
    return list(map(int, input().split()))
def get_char_list():
    s = input()
    return list(s[:len(s) - 1])
def get_tuple_ints():
    return tuple(map(int, input().split()))
def print_iterable(p):
    print(" ".join(map(str, p)))

def main():

    n, m = get_tuple_ints()
    a = get_list_ints()

    visited = [False for i in range(n)]
    g = [[] for i in range(n)]
    components = [[] for i in range(n)]

    for i in range(m):
        u, v = get_tuple_ints()
        u -= 1
        v -= 1
        g[u].append(v)
        g[v].append(u)

    component = 0
    for v in range(n):
        if visited[v]:
            continue
        stack = [v]
        while stack:
            v = stack[-1]
            stack.pop()
            if not visited[v]:
                components[component].append(v)
                for u in g[v]:
                    if not visited[u]:
                        stack.append(u)
                visited[v] = True
        component += 1

    total_components = component
    indices = sorted([i for i in range(n)], key=lambda x : -a[x])
    components.sort(key=lambda x : -len(x))

    current_rank = 0
    assigned_vertex = [0 for i in range(n)]
    for component in range(total_components):
        for x in components[component]:
            assigned_vertex[indices[current_rank]] = x + 1
            current_rank += 1

    print_iterable(assigned_vertex)
    pass

if __name__ == '__main__':
    main()
